src/utils.c

```diff
@@ -161,3 +161,41 @@ istrncmp(const char *a, const char *b, size_t n)
     }
     return 0;
 }
+
+#ifdef _MSC_VER
+#include <stdlib.h>
+
+int
+asprintf(char **strp, const char *fmt, ...)
+{
+    int r;
+    va_list ap;
+    va_start(ap, fmt);
+    r = vasprintf(strp, fmt, ap);
+    va_end(ap);
+    return r;
+}
+
+int
+vasprintf(char **strp, const char *fmt, va_list ap)
```

Comment:
It seems like MSVC does have [`vsnprintf`](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/vsnprintf-vsnprintf-vsnprintf-l-vsnwprintf-vsnwprintf-l?view=vs-2019), so `vasprintf` could be written a bit more generally like this (written quickly):

```c
int vasprintf(char **strp, const char *fmt, va_list ap) {
    int ret;
    char *buf;

    ret = vsnprintf(NULL, 0, fmt, ap);
    if (ret < 0) {
        return ret;
    }

    buf = malloc(len + 1);
    if (!buf) {
        return -1;
    }

    ret = vsnprintf(buf, ret + 1, fmt, ap);
    if (ret < 0) {
        free(buf);
        return ret;
    }

    *strp = buf;
    return ret;
}
```

What do you think about using a configure check for `vasprintf` and using the above as the fallback implementation, instead of using an MSVC-specific implementation?

Reply:
Good idea, thanks for the tip!

Reply:
The `va_list` in this case is used twice, and its value after the first call to `vsnprintf()` is undefined, so I will need to use `va_copy()` to pass a copy. Shouldn't be a problem :thinking: